keps/sig-node/3695-pod-resources-for-dra/README.md

```diff
@@ -0,0 +1,383 @@
+
+## Motivation
+
+One of the primary motivations for this KEP is to extend the PodResources API to allow node monitoring agents to access information about resources allocated by DRA. The PodResources API is also being used by CNI meta-plugins like [multus](https://github.com/k8snetworkplumbingwg/multus-cni) and [DANM](https://github.com/nokia/danm) to add resources allocated by device plugins as CNI arguments. This extension can be used to allow these CNI plugins to reference resources allocated by DRA as well. Additional extensions to the API will also make it easier for CNI meta-plugins to access resources allocated to a specific pod, rather than having to filter through resources for all pods on the node.
+
+### Goals
+
+- To allow node monitoring agents to know the allocated DRA resources for Pods on a node.
+- To allow the DRA feature to work with CNIs that require complex network devices such as RDMA. DRA resource drivers will allocate the resources, and the meta-plugin will read the allocated [CDI Devices](https://github.com/container-orchestrated-devices/container-device-interface) using the PodResources API. The meta-plugin will then inject the device-id of these CDI Devices as CNI arguments and invoke other CNIs (just as it does for devices allocated by the device plugin today).
+
+### Non-Goals
+
+To enhance the GetAllocatableResources() call in the PodResources API to account for resources managed by DRA. With DRA there is no standard way to get the capacity, for example.
+
+## Proposal
+
+### Risks and Mitigations
+
+This API is read-only, which removes a large class of risks. The aspects that we consider below are as follows:
+- What are the risks associated with the API service itself?
+- What are the risks associated with the data itself?
+
+| Risk                                                      | Impact        | Mitigation |
+| --------------------------------------------------------- | ------------- | ---------- |
+| Too many requests risk impacting the kubelet performances | High          | Implement rate limiting and or passive caching, follow best practices for gRPC resource management. |
+| Improper access to the data | Low | Server is listening on a root owned unix socket. This can be limited with proper pod security policies. |
+
+## Design Details
+
+### Proposed API
+
+
+Our proposal is to extend the existing PodResources gRPC service of the Kubelet with a repeated DynamicResource field in the ContainerResources message. This new field will contain information about the DRA resource class, the DRA resource claim and a list of CDI Devices allocated by a DRA driver.
+Additionally, we propose adding a Get method to the existing gRPC service to allow querying specific pods for their allocated resources.
+
+The extended interface is shown in proto below:
```

Comment:
I understand the proposed API is read-only and mainly for the monitoring, but not sure the entire flow. May I have a couple of questions answered before approving this KEP:

1) We are using polling here. But what if DRA driver is dead, is the pod using the resource continue running? How to make sure the stats up to date?
2) DRA might be deallocated on the node, how often kubelet to issue Get() methods to reconsolidate the status?
3) Do we need a registration logic with Kubelet? 

I guess some of the questions might be discussed in a separate KEP, but hope we have the answers to those high level questions. Thanks!

Reply:
**For (1):**
I guess you could think of it as polling, but the PodResources API is really more of an on-demand API that goes and gets the information it needs at the time it is requested. Nothing is cached as part of the API itself. The various components that hold the information (CPUManager, DeviceManager, etc) may cache this information, but the API doesn't "store" this information anywhere by itself.

In the case of the DRA manager, the information required to populate the new fields in the PodResources API is all cached in an in-memory store for each container that get launched. However, unlike the CPUManager and DeviceManager, this information is **not** persisted / checkpointed by the DRA manager itself. It **is** idempotently queriable from each DRA driver though. Meaning that the in-memory cache can be used to populate the API if the information is available, otherwise the cache can be filled by querying the DRA driver and then pushed up to the API as needed.

In response to this specific question: "But what if DRA driver is dead, is the pod using the resource continue running?"

The lifetime of a pod is not (necessarily) tied to the "readiness" of the DRA drivers that manage its resource claims. The only time a pod is blocked from deletion by a claim not being deleted (possibly because the DRA driver managing it is down) is if it is the last pod referencing that claim.

So I think what you are asking is "If a pod is running, but I can't communicate with one of the DRA drivers to query its list of DynamicResources (and this information is not available in the DRA managers in-memory cache), what do I do? Do I omit this information? Do I return an error?, etc."

Does this accurately reflect what you are trying to ask?

**For (2):**
What do you mean by "DRA might be deallocated on the node"? Do you mean the resources associated with a specific DRA driver might be deallocated on a node? And the question is how often the kubelet will need to consult with the DRA driver to refresh whatever information it may have about these resources? I think I answered this in my response to (1) when talking about the in-memory cache of the DRA manager and it ability to idempotently retrieve any missing information from the DRA driver.

The `Get()` call you refer to is something _exposed_ by the kubelet (not something the kubelet calls itself), so I assume you actually meant `NodePrepareResources()` (which is the idempotent call a kubelet can make against a DRA driver to get the list of CDI devices associated with a claim).

**For (3):**
What are you proposing should have registration logic with the kubelet? DRA resource drivers already do register with the kubelet (so it can make calls against them to Prepare/Unprepare resources). I think I need more context here to answer the question properly.

Reply:
@dchen1107
I have updated the `Proposed API` section with more details that hopefully address your comments / concerns. PTAL

https://github.com/kubernetes/enhancements/pull/3738/files#diff-190ad7c104ab2cace0e7696b41212380fa0825a6b5203ca027af4137c342e487R95

Reply:
@klueska Thanks for the detailed response.

You answered my questions. I tried to figure out how idempotent the behavior is since DRA is dynamic, the the status how persistent and consistent over the kubelet and / or DRA driver restarts. Even we might still have the inconsistency over the restart, but it is expected to some extent. I am accepting it for now. 

Also the KEP heavily depends on the DRA, but KEP doesn't include a high level component diagram capturing the dependencies and interfaces. It assumes the reviewers have all information. My original wish is to answer my questions, we could address the issue for the future readers. I guess I didn't state it clearly. I am ok with it for now.

Reply:
No problem. They were important questions to ask.

Regarding:
> Even we might still have the inconsistency over the restart, but it is expected to some extent.

I don’t think we will have any inconstancies. The PoResources API is provided by the kubelet itself, so if it is down the API is inaccessible anyway. And once it comes back up it will reconcile all of the info it needs before it starts serving the API again.

Reply:
will there be any race if API is called too early before everything got registered?

Reply:
As for plugins not responding - there will be no partial results, correct? Either all known information is returned or an error. So one plugin can be a "noisy" neighbor for the API. Should this be added as a risk?

Reply:
> will there be any race if API is called too early before everything got registered?

There shouldn't be. The set of CDI devices to allocate to a container is pulled from the various DRA plugins *before* a container is ever created. Either all plugins successfully allocate devices (thus populating the cache) and the container eventually gets created / starts running, or some plugin fails and the container start fails as well.  Meaning only containers whose cached CDI device info is complete will ever be created and thus accessible through the API.

> As for plugins not responding - there will be no partial results, correct? Either all known information is returned or an error. So one plugin can be a "noisy" neighbor for the API. Should this be added as a risk?

That shouldn't matter. If a plugin does not respond, then the container will never be started and thus not something queriable through the PodResources API.